blockdag/dag.go

```diff
@@ -1630,30 +1630,83 @@ func (dag *BlockDAG) blueBlocksBetween(lowHash, highHash *daghash.Hash, maxEntri
 	// first iterate on the selected parent chain of the highNode
 	// until we find a new highNode
 	// where highNode.blueScore-lowNode.blueScore+1 <= maxEntries
-
+	//
+	// We also keep track of highNode's selected parent chain child
+	// in case we need to calculate highNode's anticone (see below)
+	var highNodeSelectedParentChainChild *blockNode
 	for highNode.blueScore-lowNode.blueScore+1 > maxEntries {
+		highNodeSelectedParentChainChild = highNode
 		highNode = highNode.selectedParent
 	}
 
-	// Populate and return the found nodes.
-	nodes := make([]*blockNode, 0, highNode.blueScore-lowNode.blueScore+1)
-	nodes = append(nodes, highNode)
-	current := highNode
-	for current.blueScore > lowNode.blueScore {
-		for _, blue := range current.blues {
-			nodes = append(nodes, blue)
+	// Collect every node in highNode's past (including itself) but
+	// NOT in the lowNode's past (excluding itself) into an UpHeap
+	// (a heap sorted by blueScore from lowest to greatest).
+	visited := newSet()
+	highNodeAntiFuture := newUpHeap()
```

Comment:
This is actually highNodeAntiFuture∩lowNodeAntiPast.
I know it will make a very long variable name, but it's important to make this clear.

I'd rather have something generic such as `candidateBlocks` or something with comment above explaining exactly what it is, than `highNodeAntiFuture` which is plain wrong.